ocaml/forkexecd/lib/fe_stubs.c

```diff
@@ -0,0 +1,405 @@
+        while ((readed = read(pipe_fds[0], &res->msg, sizeof(res->msg))) < 0
+               && errno == EINTR)
+            continue;
+        close_fd(&pipe_fds[0]);
+        if (readed != 0 && readed < offsetof(msg_t, msg_buf) + 1) {
+            // This should never happen !!!
+            // At this point the process is created and we have a pid so
+            // we cannot just return an error.
+            // We could try to wait the process but it should fail, let
+            // returns success and let caller read process status result.
+            return 0;
+        }
+        res->msg.msg_buf[sizeof(res->msg.msg_buf) - 1] = 0;
+        if (readed > 0) {
+            // Wait the process otherwise we'll have a zombie
+            reap_pid(res->pid);
+
+            res->err_msg = res->msg.msg_buf;
+            return res->msg.err;
+        }
+    }
+    return 0;
+}
+
+CAMLprim value
+caml_safe_exec_with_helper(value args, value environment)
+{
+    CAMLparam2(args, environment);
+
+    // Copy parameters to C
+    char **c_args = copy_string_list(args);
+    char **c_envs = copy_string_list(environment);
+    if (!c_envs || !c_args) {
+        free(c_envs);
+        free(c_args);
+        caml_raise_out_of_memory();
+    }
+
+    // potentially slow section, release Ocaml engine
+    caml_enter_blocking_section();
+
+    safe_exec_result res;
+    int err = safe_exec_with_helper(&res, c_args, c_envs);
+
+    free(c_envs);
+    free(c_args);
+
+    caml_leave_blocking_section();
+
+    // error, notify with an exception
+    if (err != 0)
+        unix_error(err, res.err_msg, Nothing);
+
+    CAMLreturn(Val_int(res.pid));
+}
+
+CAMLprim value
+caml_pidwaiter_dontwait(value pid_val)
+{
+    CAMLparam1(pid_val);
+    pid_t pid = Int_val(pid_val);
+
+    // reap the pid to avoid zombies
+    pthread_t th;
+    if (create_thread_minstack(&th, thread_proc_reap, (void *) (intptr_t) pid) == 0)
+        pthread_detach(th);
+
+    CAMLreturn(Val_unit);
+}
+
+typedef struct {
+    pid_t pid;
+    bool timed_out;
+    bool stop;
+    struct timespec deadline;
+    pthread_mutex_t mtx;
+    pthread_cond_t cond;
+} timeout_kill;
+
+static void *
+thread_proc_timeout_kill(void *arg)
+{
+    timeout_kill *tm = (timeout_kill *) arg;
+
+    pthread_mutex_lock(&tm->mtx);
+    int res  = tm->stop ? 0:
+        pthread_cond_timedwait(&tm->cond, &tm->mtx, &tm->deadline);
+    pthread_mutex_unlock(&tm->mtx);
+
+    if (res == ETIMEDOUT) {
+        kill(tm->pid, SIGKILL);
+        tm->timed_out = true;
+    }
+    return NULL;
+}
+
+static int
+create_thread_minstack(pthread_t *th, void *(*proc)(void *), void *arg)
+{
+    int res;
+
+    // disable any possible signal handler so we can safely use a small stack
+    // for the thread
+    sigset_t sigset, old_sigset;
+    sigfillset(&sigset);
+    pthread_sigmask(SIG_BLOCK, &sigset, &old_sigset);
+
+    pthread_attr_t th_attr;
+    res = pthread_attr_init(&th_attr);
+    if (res)
+        return res;
```

Comment:
if this returns early sigmask remains blocked. 
Something like this could be used to have a try/finally like wrapper that ensures we always restore the mask.
```
pthread_sigmask(SIG_BLOCK, &sigset, &old_sigset);
do {
  ...
  if (res)
    break;
} while(0);
pthread_sigmask(SIG_SETMASK, &old_sigset, NULL);
```

Reply:
fixed